### PR TITLE
feat: Add contract astroport-pair-xyk-sale-tax

### PR DESCRIPTION
**Pull request by apollo-sturdy**
This contract is a copy of the standard XYK pair but adds a configurable tax on swaps.

After launching the APOLLO token, Apollo DAO will be submitting a proposal to migrate standard xyk pair to this contract. If the proposal passes, perhaps it makes sense to merge it into the main astroport-core repo.

### CODE REVIEW

**Review by pacmanifold — APPROVED** [review]

File: contracts/pair_xyk_sale_tax/src/contract.rs [hunk 1]
```diff
@@ -1092,15 +1126,28 @@ pub fn query_asset_balances_at(
...
+/// * **tax_config** tax configuration for the swap.
 pub fn compute_swap(
     offer_pool: Uint128,
     ask_pool: Uint128,
-    offer_amount: Uint128,
+    offer_asset: &Asset,
     commission_rate: Decimal,
-) -> StdResult<(Uint128, Uint128, Uint128)> {
+    tax_config: Option<&TaxConfigChecked>,
+) -> StdResult<(Uint128, Uint128, Uint128, Uint128, Uint128)> {
```
[the remark this answers is not in the document]
Comment (apollo-sturdy) [reply contracts/pair_xyk_sale_tax/src/contract.rs]: Fixed.

**Comment by apollo-sturdy** [comment]
> Hi @apollo-sturdy
> 
> Isn't this recently added feature does the same thing?
> 

Thanks for mentioning this, I wasn't aware of this change. However, this doesn't seem to allow us to accomplish what we want. There are a few differences:

- It charges the fee share / tax on the return asset rather than the offer asset.
- It charges the fee rate on top of the commission amount which is set the same for all pairs of the same type, which means we at most would be able to take 100% of the `total_fee_rate`.

**Comment by donovansolms** [comment]
Looks good - I have nothing to add. 

I hope that `tax_config_admin` would be some governance address and not a multisig that would allow silent rugging of everyone that swaps via this pool. If it is, I'd much rather have stricter limits on what the tax can be set to.

File: contracts/pair_xyk_sale_tax/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+[package]
+name = "astroport-pair-xyk-sale-tax"
+version = "1.4.0"
```
Comment (donovansolms) [inline contracts/pair_xyk_sale_tax/Cargo.toml:3]: I think this should be bumped to 1.4.1

File: contracts/pair_xyk_sale_tax/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,1426 @@
...
+    }
+
+    let init_params = SaleTaxInitParams::from_binary(msg.init_params.clone())?;
+
+    let config = Config {
+        pair_info: PairInfo {
+            contract_addr: env.contract.address.clone(),
+            liquidity_token: Addr::unchecked(""),
+            asset_infos: msg.asset_infos.clone(),
+            pair_type: PairType::Custom(CONTRACT_NAME.to_string()),
```
Comment (epanchee) [inline contracts/pair_xyk_sale_tax/src/contract.rs:72]: As Astroport is not going to add this specific pair type in the factory query to get fee config from factory will always fail. I suggest to keep `PairType::Xyk {}` for this pair.


The other path would be to add config for this pair with `is_disabled: true` but this blocks this pair from being incentivized in [Generator](https://github.com/astroport-fi/astroport-core/blob/1bfab32a56aedcf341628b70e457da2687be34a9/contracts/tokenomics/generator/src/contract.rs#L2011-L2013). Don't know why we have this check `pair_config.is_disabled` in generator tho.

**Review by epanchee — CHANGES_REQUESTED** [review]

File: contracts/pair_xyk_sale_tax/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,1426 @@
...
+            maker_fee_amount = f.amount;
+            messages.push(f.into_msg(fee_address)?);
+        }
+    }
+
+    if config.track_asset_balances {
+        BALANCES.save(
+            deps.storage,
+            &offer_pool.info,
+            &(offer_pool.amount + offer_amount),
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/pair_xyk_sale_tax/src/contract.rs:727]: Sorry my bad missed the tax is subtracted in `compute_swap`. Resolved

File: contracts/pair_xyk_sale_tax/Cargo.toml (hunk 2 above)
Comment (apollo-sturdy) [reply contracts/pair_xyk_sale_tax/Cargo.toml:3]: Or set back to 1.0.0 as it's a separate contract from the standard xyk? Doesn't matter either way to me, I can change it to 1.4.1 if you prefer.

File: contracts/pair_xyk_sale_tax/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+[features]
+# for quicker tests, cargo test --lib
+# for more explicit tests, cargo test --features=backtraces
+backtraces = ["cosmwasm-std/backtraces"]
+library = []
+
+[dependencies]
+integer-sqrt = "0.1"
+astroport = { path = "../../packages/astroport", default-features = false }
```
[the remark this answers is not in the document]
Comment (apollo-sturdy) [reply contracts/pair_xyk_sale_tax/Cargo.toml:28]: Good point. I guess this is blocked by having a up to date (with only those changes that are live on mainnet) version of astroport package published on crates.io.

File: contracts/pair_xyk_sale_tax/src/contract.rs [hunk 6]
```diff
@@ -0,0 +1,1426 @@
...
+            block_time_last: old_config.block_time_last,
+            pair_info: old_config.pair_info,
+            price0_cumulative_last: old_config.price0_cumulative_last,
+            price1_cumulative_last: old_config.price1_cumulative_last,
+            track_asset_balances: old_config.track_asset_balances,
+        };
+        CONFIG.save(deps.storage, &new_config)?;
+    }
+
+    Ok(Response::default())
```
[the remark this answers is not in the document]
Comment (apollo-sturdy) [reply contracts/pair_xyk_sale_tax/src/contract.rs]: Good call, thanks. Fixed.

File: contracts/pair_xyk_sale_tax/src/contract.rs (hunk 3 above)
Comment (apollo-sturdy) [reply contracts/pair_xyk_sale_tax/src/contract.rs:72]: Ah yeah this was from my original idea to add ut as a custom pair type until speaking with @stefanionescu and he suggested that we should instead migrate the standard xyk pair. So in practice this contract will never be instantiated (unless someone wants to instantiate an instance outside of astroport). I can change it to `PairType::Xyk {}` if you prefer.

Comment (apollo-sturdy) [reply contracts/pair_xyk_sale_tax/src/contract.rs]: changed it.

File: contracts/pair_xyk_sale_tax/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/pair_xyk_sale_tax/Cargo.toml:28]: True. We'll deal with it once this issue resolved

File: contracts/pair_xyk_sale_tax/src/contract.rs [hunk 7]
```diff
@@ -0,0 +1,1442 @@
...
+use astroport::querier::{query_factory_config, query_fee_info, query_supply};
+use astroport::{token::InstantiateMsg as TokenInstantiateMsg, U256};
+use cw_utils::parse_instantiate_response_data;
+
+use crate::error::ContractError;
+use crate::state::{Config, BALANCES, CONFIG};
+use astroport::pair_xyk_sale_tax::{
+    MigrateMsg, SaleTaxConfigUpdates, SaleTaxInitParams, TaxConfigChecked,
+};
+use astroport_pair::state::{Config as XykConfig, CONFIG as XYK_CONFIG};
```
Comment (epanchee) [inline contracts/pair_xyk_sale_tax/src/contract.rs:36]: Could you confirm the astroport-pair versions you support in migration could be deserialized with these structures?

Comment (apollo-sturdy) [reply contracts/pair_xyk_sale_tax/src/contract.rs:36]: I added a test migrating from 1.3.1. Unfortunately 1.3.0 and 1.4.0 are not on crates.io. Can you add them?

**Review by epanchee — APPROVED** [review]

**Review by epanchee — APPROVED** [review]
LGTM

**Pull request merged**